ext-dynamic-commitments.md

```diff
@@ -0,0 +1,1087 @@
+constraints on these parameters to make sure they are internally consistent and
+secure in all contexts.
+
+Since the `initiator` is the one that is responsible for paying the fees for the
+kickoff transaction if it is required (like for certain `channel_type` changes),
+it follows that the `responder` cannot change the `channel_type`. Since the
+`kickoff_feerate` is paid by the `initiator`, it should be set only if the
+sender is the `initiator`.
+
+The requirement for a node to remember what it last _sent_ and for it to
+remember what it _accepted_ is necessary to recover on reestablish. See the
+reestablish section for more details.
+
+#### `dyn_ack`
+
+This message is sent in response to a `dyn_propose` indicating that it has
+accepted the proposal.
+
+1. type: 113 (`dyn_ack`)
+2. data:
+   * [`32*byte`:`channel_id`]
+
+##### Requirements
+
+The sending node:
+  - MUST set `channel_id` to a valid channel it has with the recipient.
+  - MUST NOT send this message if it has not received a `dyn_propose` for this
+    `channel_id`
+  - MUST NOT send this message if it has already sent a `dyn_ack` for the
+    current negotiation.
+  - MUST NOT send this message if it has already sent a `dyn_reject` for the
+    current negotiation.
+  - MUST remember the parameters of `dyn_propose` message to which the `dyn_ack`
+    is responding for the next `dyn_height`. (See `channel_reestablish`
+    requirements)
+  - MUST remember the local and remote commitment heights for the next
+    `dyn_height`.
+
+The receiving node:
+  - if `channel_id` does not match an existing channel it has with the peer:
+    - MUST send an `error` and close the connection.
+  - if there isn't an outstanding `dyn_propose` it has sent:
+    - MUST send an `error` and fail the channel.
+
+A node:
+  - once it has sent or received `dyn_ack`
+    - MUST increment its `dyn_height`.
+    - MUST proceed to the Execution Phase.
+
+##### Rationale
+
+The `dyn_height` starts at 0 for a channel and is incremented by 1 every
+time the dynamic commitment proposal phase completes for a channel. See the
+reestablish section for why this is needed.
+
+#### `dyn_reject`
+
+This message is sent in response to a `dyn_propose` indicating that it rejects
+the proposal.
+
+1. type: 115 (`dyn_reject`)
+2. data:
+    * [`32*byte`:`channel_id`]
+    * [`...*byte`:`update_rejections`]
+
+##### Requirements
+
+The sending node:
+  - MUST set `channel_id` to a valid channel it has with the recipient.
+  - MUST NOT send this message if it has not received a `dyn_propose`
+  - MUST NOT send this message if it has already sent a `dyn_ack` for the
+    current negotiation.
+  - MUST NOT send this message if it has already sent a `dyn_reject` for the
+    current negotiation.
+  - if it will not accept **any** dynamic commitment negotiation:
+    - SHOULD send a `dyn_reject` with zero value for `update_rejections`
+  - if it does not agree with one or more parameters:
```

Comment:
The idea of odd/even fields doesn't really map cleanly to a scenario where we actively need to reject fields (ie, we can't just ignore odd values like we do elsewhere).

This doesn't _feel_ like a place where we'll want to extend functionality (as there are only so many channel params), but for the sake of future-proofing WDYT of something like:
* Set current negotiation TLVs to even values
* Reject any unknown odd TLVs in `update_rejections`

That provides a possible upgrade path where we can start adding new values with odd tlv field numbers and know whether our counterparty actually understood them or not.

Reply:
> This doesn't feel like a place where we'll want to extend functionality

This is proooooobably right. Although given we already have `open_channel` and `open_channel2` I was going about this a bit more defensively.

> Set current negotiation TLVs to even values

Definitely agree here.

> Reject any unknown odd TLVs in `update_rejections`

Yeah rejection is purely discretionary and this field is largely here to improve the overall UX. We could always provide unstructured or no information on why a proposal was rejected but I figured with a strictly enumerable set of parameters that marking the ones we find dissatisfactory would allow progress to be made more effectively.

As an aside, I don't know that legitimate use of odd fields here will ever make sense, since the whole purpose here is to coordinate on fundamental values. This careful coordination is generally incompatible with the idea of safely ignoring unknown values, so I think your assessment is correct that it will go unused. Even still, I think your recommendation makes sense because even if we never use them I think state machines negotiating the changes should reject things they don't understand in this context.